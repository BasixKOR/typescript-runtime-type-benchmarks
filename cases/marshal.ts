import 'reflect-metadata'
import { f, validatedPlainToClass, PropertyValidator, PropertyValidatorError } from '@marcj/marshal'
import { Data } from '../data'

class IsNegative implements PropertyValidator {
  validate<T>(value: number) {
    if (value > 0) {
      return new PropertyValidatorError('IsNegative', 'Number must be negative.')
    }
  }
}

function MinLengthFactory (minLength: number) {
  return class MinLength implements PropertyValidator {
    validate<T>(value: string) {
      if (value.length < minLength) {
        return new PropertyValidatorError('MinLength', `String must have minimum length of ${minLength}.`)
      }
    }
  }
}

type DeeplyNested = Data['deeplyNested']

class DeeplyNestedType implements DeeplyNested {
  @f
  foo!: string

  @f
  num!: number

  @f
  bool!: boolean
}

class DataType implements Data {
  @f
  number!: number

  @f.validator(IsNegative)
  neg_number!: number

  @f
  max_number!: number

  @f
  string!: string

  @f.validator(MinLengthFactory(100))
  long_string!: string

  @f
  boolean!: boolean

  @f.type(DeeplyNestedType)
  deeplyNested!: DeeplyNestedType
}

export function caseMashal(data: Data) {
  return validatedPlainToClass(DataType, data)
}
