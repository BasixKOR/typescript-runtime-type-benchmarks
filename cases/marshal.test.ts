import { caseMashal } from './marshal'
import { DATA } from '../data'

describe('caseMashal', () => {
  it('should pass', () => {
    const res = caseMashal(DATA)
    expect(res).toEqual(DATA)

    if (res) {
      expect(res.number).toEqual(DATA.number)
    }
  })

  it('should fail negative number test', () => {
    const req = Object.assign({}, DATA, { neg_number: 1 })
    expect(() => caseMashal(req)).toThrowErrorMatchingSnapshot()
  })

  it('should fail min length test', () => {
    const req = Object.assign({}, DATA, { long_string: '' })
    expect(() => caseMashal(req)).toThrowErrorMatchingSnapshot()
  })
})