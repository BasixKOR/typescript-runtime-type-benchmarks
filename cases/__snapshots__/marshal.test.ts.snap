// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`caseMashal should fail min length test 1`] = `undefined`;

exports[`caseMashal should fail negative number test 1`] = `undefined`;
